alma/alma.go

```diff
@@ -0,0 +1,161 @@
+	Version      string      `json:"version"`
+}
+
+type OID struct {
+	OID string `json:"$oid,omitempty"`
+}
+
+type Date struct {
+	Date int64 `json:"$date"`
+}
+
+type Pkglist struct {
+	Name      string    `json:"name"`
+	Shortname string    `json:"shortname"`
+	Packages  []Package `json:"packages"`
+}
+
+type Package struct {
+	Name            string      `json:"name"`
+	Version         string      `json:"version"`
+	Release         string      `json:"release"`
+	Epoch           string      `json:"epoch"`
+	Arch            string      `json:"arch"`
+	Src             string      `json:"src"`
+	Filename        string      `json:"filename"`
+	Sum             string      `json:"sum"`
+	SumType         interface{} `json:"sum_type"`
+	RebootSuggested int         `json:"reboot_suggested"`
+}
+
+type Module struct {
+	Stream  string `json:"stream,omitempty"`
+	Name    string `json:"name,omitempty"`
+	Version int64  `json:"version,omitempty"`
+	Arch    string `json:"arch,omitempty"`
+	Context string `json:"context,omitempty"`
+}
+
+type Reference struct {
+	Href  string `json:"href"`
+	Type  string `json:"type"`
+	ID    string `json:"id"`
+	Title string `json:"title"`
+}
+
+type Config struct {
+	VulnListDir string
+	URLs        map[string]string
+	Retry       int
+}
+
+func NewConfig() Config {
+	urls := map[string]string{}
+	for version := range AlmaReleaseVersion {
+		urls[version] = fmt.Sprintf(urlFormat, version)
+	}
+
+	return Config{
+		VulnListDir: utils.VulnListDir(),
+		URLs:        urls,
+		Retry:       retry,
+	}
+}
+
+func (c Config) Update() error {
+	for version, url := range c.URLs {
+		log.Printf("Fetching security advisories of AlmaLinux %s ...\n", version)
+		if err := c.update(version, url); err != nil {
+			return xerrors.Errorf("failed to update security advisories of AlmaLinux %s: %w", version, err)
+		}
+	}
+	return nil
+}
+
+func (c Config) update(version, url string) error {
+	dirPath := filepath.Join(c.VulnListDir, almaLinuxDir, version)
+	log.Printf("Remove AlmaLinux %s directory %s\n", version, dirPath)
+	if err := os.RemoveAll(dirPath); err != nil {
+		return xerrors.Errorf("failed to remove AlmaLinux %s directory: %w", version, err)
+	}
+	if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
+		return xerrors.Errorf("failed to mkdir: %w", err)
+	}
+
+	body, err := utils.FetchURL(url, "", c.Retry)
+	if err != nil {
+		return xerrors.Errorf("failed to fetch security advisories from AlmaLinux: %w", err)
+	}
+
+	var errata []erratum
+	if err := json.Unmarshal(body, &errata); err != nil {
+		return xerrors.Errorf("failed to unmarshal json: %w", err)
+	}
+
+	var secErrata []erratum
+	for _, erratum := range errata {
+		if !strings.HasPrefix(erratum.UpdateinfoID, "ALSA-") {
+			continue
+		}
+		secErrata = append(secErrata, erratum)
+	}
+
+	bar := pb.StartNew(len(secErrata))
+	for _, erratum := range secErrata {
+		filepath := filepath.Join(dirPath, fmt.Sprintf("%s.json", erratum.UpdateinfoID))
```

Comment:
Can we have one more directory per year?
https://github.com/aquasecurity/vuln-list/tree/main/redhat

Sometimes I only want to see the vulnerabilities of the year locally, and it's slow if all the files are in the same directory.

Reply:
Currently, JSON is output like this under `vuln-list/alma/8`.
```console
$ tree ~/.cache/vuln-list-update/vuln-list/alma/8
/home/mainek00n/.cache/vuln-list-update/vuln-list/alma/8
├── ALSA-2019:1152.json
├── ALSA-2020:0279.json
└── ALSA-2021:0150.json
```

Is it correct that I understand that you want to divide this by year as follows?
```console
$ tree ~/.cache/vuln-list-update/vuln-list/alma/8
/home/mainek00n/.cache/vuln-list-update/vuln-list/alma/8
├── 2019
│   └── ALSA-2019:1152.json
├── 2020
│   └── ALSA-2020:0279.json
└── 2021
    └── ALSA-2021:0150.json
```

Reply:
Yes, looks good!